x/programs/rust/wasmlanche-sdk/src/state.rs

```diff
@@ -84,60 +97,81 @@ impl<'a, K: Key> State<'a, K> {
     /// the host fails to read the key and value.
-    /// # Panics
-    /// Panics if the value cannot be converted from i32 to usize.
     pub fn get<V>(self, key: K) -> Result<Option<V>, Error>
     where
         V: BorshDeserialize,
     {
-        #[link(wasm_import_module = "state")]
-        extern "C" {
-            #[link_name = "get"]
-            fn get_bytes(ptr: *const u8, len: usize) -> HostPtr;
+        match self.cache.borrow_mut().entry(key) {
+            Entry::Occupied(entry) => match entry.get() {
+                CachedData::Data(data) => from_slice::<V>(data)
+                    .map_err(|_| StateError::Deserialization)
+                    .map(Some),
+                CachedData::ToDelete | CachedData::Deleted => Ok(None),
+            },
+            Entry::Vacant(entry) => {
+                if let Some(bytes) = Self::get_from_host(key)? {
+                    let value = from_slice::<V>(&bytes)
+                        .map_err(|_| StateError::Deserialization)
+                        .map(Some);
+                    entry.insert(CachedData::Data(bytes));
+                    value
+                } else {
+                    entry.insert(CachedData::Deleted);
+                    Ok(None)
+                }
+            }
         }
+    }
 
-        let mut cache = self.cache.borrow_mut();
-
-        let val_bytes = if let Some(val) = cache.get(&key) {
-            val
-        } else {
-            let args_bytes = borsh::to_vec(&key).map_err(|_| StateError::Serialization)?;
-
-            let ptr = unsafe { get_bytes(args_bytes.as_ptr(), args_bytes.len()) };
-
-            if ptr.is_null() {
-                return Ok(None);
+    /// # Errors
+    /// Returns an [Error] if the key cannot be serialized,
+    /// if the host fails to handle the operation,
+    /// or if the bytes cannot be borsh deserialized.
+    pub fn delete<V>(self, key: K) -> Result<Option<V>, Error>
+    where
+        V: BorshDeserialize,
+    {
+        match self.cache.borrow_mut().entry(key) {
+            Entry::Occupied(mut entry) => match entry.get_mut() {
+                CachedData::Data(data) => {
+                    let value = from_slice::<V>(data)
+                        .map_err(|_| StateError::Deserialization)
+                        .map(Some);
+                    entry.insert(CachedData::Deleted);
+                    value
+                }
+                CachedData::Deleted | CachedData::ToDelete => Ok(None),
+            },
+            Entry::Vacant(entry) => {
+                if let Some(data) = Self::get_from_host(key)? {
+                    entry.insert(CachedData::ToDelete);
+                    from_slice::<V>(&data)
+                        .map_err(|_| StateError::Deserialization)
+                        .map(Some)
+                } else {
+                    entry.insert(CachedData::Deleted);
+                    Ok(None)
+                }
```

Comment:
Same as the comment above, I don't think you need all the repetition. There's a lot of surface area for bugs here.